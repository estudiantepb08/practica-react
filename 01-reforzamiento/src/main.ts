import './style.css';
import './bases/01-const-let';
import './bases/02-template-string';
import './bases/03-object-literal.ts';
import './bases/04-arrays.ts';
import './bases/05-funtions.ts';
import './bases/06-obj-destructuring.ts';
import './bases/07-array-destructuring.ts';
import { getHeroeByOwner } from './bases/08-imp-exp.ts'
import { Owner } from './data/heroes.data.ts';
import './bases/09-promises.ts';
import './bases/10-fetch-api.ts';
import './bases/11-async-await.ts';

document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
  <div>
    <h1>HOLA MUNDO</h1>
  </div>
`
console.log(getHeroeByOwner(Owner.Marvel));
console.log(getHeroeByOwner(Owner.DC));

// Importante: Si se importa algo que no se usa, da error. Por eso se importan los otros archivos

// Ejemplos de desestructuraci√≥n


