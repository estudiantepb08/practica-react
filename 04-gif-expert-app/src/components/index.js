/**Es un archivo de barril para utilizar todos los componentes y hacer una solo importacion como lo hace react */

export * from './AddCategories';
export * from './GifGrid';
export * from './GifItem';