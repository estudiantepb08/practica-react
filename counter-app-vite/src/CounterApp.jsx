import React from 'react';
import { useState } from 'react';

export const CounterApp = ({value})=>{
    /**
     * Creamos nuestro primer hook que es el estado el hook renderiza o dibuja el nuevo estado del componente
     * utiliza los argumentos el valor que va a dibujar y donde lo voy a setear con set()
     * siempre deben ser una constante
     * const [counter, setCounter] = useState(0);
     */
    const [counter, setCounter] = useState(value);

    const handleApp = () =>{setCounter(counter + 1)};
    const handleSubtract = () =>{setCounter(counter - 1)};
    const handleReset = () =>{setCounter(value)};
    /*
    se puede enviar solo la referencia a la funcion sin enviarle la cordenada completa
    tradicional
    <button onClick={(event) => handleApp(event)}> + 1</button>
    solo referencia en la funcion se pasan los argumentos de la misma forma
    <button onClick={handleApp}> + 1</button>
    */
   /**
    * Para realizar las pruebas unitarias necesitamos installar en VITE la libreria de JEST
    * Y en el package.json agregar la propiedad "test":"jest" "test": "jest --watchAll"
    * para ejecutar las pruebas es con el comando yarn test o npm run test 
    * otra manera de identificar un elemento HTML es con aria-label
    */

    return(
        <>
        <h1>Evento clic</h1>
        <h2>{ counter }</h2>
        <button onClick={ handleApp }> + 1</button>
        <button onClick={ handleSubtract }> - 1</button>
        <button aria-label='btn-reset' onClick={ handleReset }> Reset</button>
        </>
    );
}
