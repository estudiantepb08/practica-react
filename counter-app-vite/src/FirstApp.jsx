import React from 'react';

const newMessages = 'ALEX'; 
/*
esta constantes deben inr fuera de la funcion para que no se ha renderizado
No he permitodo los objetos json, solo es permitido serializandolo con la clase JSON.stringify
*/

const getResultado = (a,b) =>{
    return a + b;
};

export const FirstApp = () =>{
    return (
        <>
    <h2>First App</h2>
    <p>hola: { newMessages }</p>
    <p>Valor Numero: {getResultado(2,8)}</p>
    </>
    );
};