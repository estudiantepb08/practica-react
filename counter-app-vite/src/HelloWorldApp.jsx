import React from 'react';

export function App(){
    return (<h1>Hellow Work</h1>)
}

export default App;
