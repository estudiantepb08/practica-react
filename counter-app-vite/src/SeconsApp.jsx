import React from 'react';
import PropTypes from 'prop-types';

export const SeconsApp = ({title, subtitle}) =>{
    /**
     * Las props se pueden desestructurar colocandolas entre llaves {} o solas
     * es decir el nombre del objecto dentro del argumento del componente
     */
    return (<>
    <h1 data-testid="test-title"> { title } </h1>
    <p>Mis primeras props titulo: Trabajando Con Las Props O Propiedades De La Aplicaci√≥n:</p>
    <p>Mis primeras props subtitulo: { subtitle +1}</p>
    </>);
};

/**
 * Aqui se deben definir las propiedades que va a utilizar el componente
 * se define el tipo de variable, si es obligatoria con la propiedad .isRequired
 */
SeconsApp.propTypes = {
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.number.isRequired
};
/**
 * Tambien podemos definir propiedades por defecto con el props * 
 */

SeconsApp.defaultProps ={
    title: 'Ejemplo No hay tiulo',
    /*subtitle: 23*/
}