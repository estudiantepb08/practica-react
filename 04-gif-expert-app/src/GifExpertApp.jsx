import React from 'react';

import { useState } from "react";
import { AddCategories, GifGrid } from './components';

export const GifExpertApp = ()=>{

    const [categories, setCategories] = useState(['One Punch','Saludo']); 
    /** El valor inicial que pase en el useState lo obtengo con el segundo argumento es decir
     * categories seria mi getCategories, y setCategories me recibe el valor
     */

    const onAddCategory = (newCategories)=>{
        /**Validamos que la categoria no este repetidad en el arreglo */
        if(categories.includes(newCategories)) return;
        //categories.push('Que tal');
        setCategories([newCategories, ...categories]); // hacer copia de categoria y agregar el nuevo valor 'Forma express'
        //setCategories(cat => [...cat, 'valorant']); // con callback
//<button onClick={onAddCategory}>Agregar</button>
    };

    return(<>
    {/** Titulo */}
    <h1>GitExpertApp</h1>
    {/** Input */}
    <AddCategories onNewCategories={onAddCategory}></AddCategories>
    {/** Listado de Gif */}
    
    <ol>{
       /* Primera forma de poder listar la lista de categorias
        categories.map(category =>{
            return<li key={category}>{category}</li>
        })
        */
       /** Segunda forma de listar o recorrer la lista desestructurando con un componente */

       categories.map(category =>(
        <GifGrid key={category} category={category}></GifGrid> //Componente GifGrid categori para desacoplar el codigo
       ))
        }
        
        {/*<li>123</li>
        <li>XYZ</li>*/}
    </ol>
    {/** Gif Item */}
    </>);
}