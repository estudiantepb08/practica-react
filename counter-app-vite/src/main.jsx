import ReactDOM from 'react-dom'
import React from 'react'
import { createRoot } from 'react-dom/client';
import App from './HelloWorldApp';

import moduleName from 'module'
import { FirstApp } from './FirstApp';
import './styles.css'; // importar hoja de estilos de manera global para todos los componentes
import { SeconsApp } from './SeconsApp';
import { CounterApp } from './CounterApp';

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        {/*<App></App>
        <FirstApp></FirstApp>
        <SeconsApp title='PROPS123' subtitle={12}></SeconsApp>
        */
        }
        
        <CounterApp value={2}></CounterApp>
        
    </React.StrictMode>
);
