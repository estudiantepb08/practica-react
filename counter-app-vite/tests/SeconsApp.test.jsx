import { render, screen } from "@testing-library/react";
import { SeconsApp } from "../src/SeconsApp";
/**
 * se deben de configurar los archivos
 * jest.config.cjs
 * jest.setup.js
 * babel.config.cjs
 * y instalar jsdom
 * el render actualiza el objeto scrint
 * el container, se parece a un nodo de html. crea una carpeta llamada snapshot con el componente que se esta probando
 * en las pruebas se puede utilizar javaScript
 * El snapshots, hace que la prueba se ha rigida, si se modifica los elementos del componente se afecta la prueba
 * porque el toma la primera foto.
 * se pueden utitilizar otras funciones del render para validar los elementos.
 * se puede utilizar constantes cuando hay variables repetidas
 * para no utiliar el container se puede utilizar la importacion del screen
 * import { render, screen } from "@testing-library/react";
 * 
 * npm install --save-dev @testing-library/react
 * npm install --save-dev @testing-library/react
 * npm i jest-environment-jsdom-global
 * npm i whatwg-fetch
 * npm i prop-types
 * https://www.npmjs.com/package/prop-types
 */
describe('Pruebas del componente <SeconsApp/>', ()=>{

const title2 = 'PROP123';

    test('debe mostrar el titulo en h1', () =>{

        const title = 'PROPS123';
        const {container, getByText, getByTestId} = render(<SeconsApp title = {title}/>);
        expect(getByText(title)).toBeTruthy();

        /* con snapshots
        const h1 = container.querySelector('h1');
        expect(h1.innerHTML).toContain(title);
        */
       /** Sin snapshots solo con jest y render el toBe() es estricto debe cumplirse todo
        * el toContainer() es que contenga el elemento
        */
       expect(getByTestId('test-title').innerHTML).toContain(title);


    });

    test('Pruebas SeconsApp debe hacer match con el snapshot', ()=>{

        const subtitle = 12;        
        const {container} = render(<SeconsApp subtitle = {subtitle}/>);
        expect(container).toMatchSnapshot();
        
    });

    /** Otra opcion sin el container es con el screen tambien se puede buscar por getByRole() los elementos */

    test('Debe mostrar un mensaje PROP123', ()=>{
        render(<SeconsApp title= {title2}></SeconsApp>);
        expect(screen.getByText(title2)).toBeTruthy();
    });

});