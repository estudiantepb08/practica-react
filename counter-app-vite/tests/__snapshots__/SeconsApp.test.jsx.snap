// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <SeconsApp/> Pruebas SeconsApp debe hacer match con el snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
     
    Ejemplo No hay tiulo
     
  </h1>
  <p>
    Mis primeras props titulo: Trabajando Con Las Props O Propiedades De La Aplicaci√≥n:
  </p>
  <p>
    Mis primeras props subtitulo: 
    13
  </p>
</div>
`;
